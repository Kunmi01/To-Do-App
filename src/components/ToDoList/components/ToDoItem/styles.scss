@import '../../../../scss/base.scss';

.todo-item {
  background: $background-dark;
  border-radius: $unit;
  margin-bottom: $unit * 5;

  &:last-child {
    margin-bottom: 0;
  }

  &:not(.todo-item--creation) {
    overflow: hidden;
    animation: expand_down 0.33s ease-in 0s 1 normal;
    @keyframes expand_down {
      from {
        opacity: 0;
        max-height: 0;
      }
      to {
        opacity: 1;
        max-height: 500px;
      }
    }
  }

  &--editable {
    background-color: $grey2;

    .todo-item__form {
      &__name,
      &__description {
        background-color: $white;
        padding: $unit $unit * 2;
      }
    }
  }

  &--creation {
    background-color: $light-grey;
    margin-bottom: $unit * 12;
  }

  &__form {
    display: flex;
    flex-direction: column;
    padding: $unit * 4;

    &__name,
    &__description {
      border: 1px solid $grey;
      border-radius: $unit;
      background-color: inherit;
      font-size: $font-m;
      margin-bottom: $unit * 2;

      &:focus {
        outline: none;
        border-color: $deep-green;
      }

      &:disabled {
        color: $text-light;
        cursor: not-allowed;
      }

      &::placeholder {
        color: $grey;
      }
    }

    &__name {
      font-family: $font-play-bold;
      font-size: $font-ml;

      &:disabled {
        border: 0;
      }
    }

    &__description {
      font-family: $font-play-regular;
      min-height: $unit * 32;
      max-height: $unit * 40;
      padding: $unit * 2;
      overflow: scroll;
      resize: vertical;

      &:disabled {
        border: 1px solid $grey2;
        font-size: $font-s;
      }

      @include mq-from('mobile') {
        min-height: $unit * 22;
      }

      @include mq-from('tablet') {
        min-height: $unit * 12;
      }
    }

    &__bottom-wrapper {
      display: flex;
      flex-direction: column;

      @include mq-from('tablet') {
        flex-direction: row;
        align-items: center;
      }
    }

    &__buttons-wrapper {
      display: flex;
      margin-left: auto;
      margin-top: $unit * 4;

      @include mq-from('tablet') {
        margin-top: 0;
      }
    }

    &__button {
      border: 0;
      border-radius: $unit;
      font-family: $font-play-bold;
      font-size: $font-s;
      color: $text-light;
      opacity: 0.88;
      padding: $unit * 2 $unit * 4;
      margin-left: $unit * 4;
      cursor: pointer;

      &:hover {
        opacity: 1;
      }

      &:disabled {
        color: $light-grey;
        opacity: 0.35;
        cursor: not-allowed;
      }

      &--submit {
        background-color: $deep-green;
        margin-left: auto;
      }

      &--edit {
        background-color: $grey2;
      }

      &--delete {
        background-color: $dark-red;
      }
    }

    &__created {
      color: $grey;
      font-size: $font-xs;
    }
  }
}
