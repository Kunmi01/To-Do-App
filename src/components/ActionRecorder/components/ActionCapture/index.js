import React from 'react';
import PropTypes from 'prop-types';
import './styles.scss';

const ActionCapture = ({ type, payload }) => (
  <div className="action-capture">
    <div className="action-capture__type">{type}</div>
    <div className="action-capture__payload">
      {JSON.stringify(payload).split('"').join(' ')}
    </div>
  </div>
);

ActionCapture.propTypes = {
  type: PropTypes.string.isRequired,
  payload: PropTypes.shape({
    itemId: PropTypes.string.isRequired,
    itemName: PropTypes.string,
    itemDescription: PropTypes.string,
    itemCreatedDate: PropTypes.string
  }).isRequired
};

export default ActionCapture;
