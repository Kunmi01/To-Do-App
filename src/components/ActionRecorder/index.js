import React from 'react';
import './styles.scss';

const ActionRecorder = () => {
  return (
    <div className="action-recorder">
      <div className="action-recorder__wrapper">
        <h2 className="action-recorder__header">Action Recorder</h2>
        <div className="action-recorder__inner-wrapper">Content here</div>
      </div>
    </div>
  );
};

export default ActionRecorder;
