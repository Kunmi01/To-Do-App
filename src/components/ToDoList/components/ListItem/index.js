import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import './styles.scss';
import { TODO_LIST_ITEMS } from '../../../../utils/constants';

const shortid = require('shortid');

const ListItem = ({
  editable,
  creation,
  id,
  name,
  description,
  createdDate
}) => {
  const [itemEditable, setItemEditable] = useState(false);
  const [itemId, setItemId] = useState('');
  const [itemName, setItemName] = useState('');
  const [itemDescription, setItemDescription] = useState('');
  const [itemCreatedDate, setItemCreatedDate] = useState('');

  useEffect(() => {
    if (editable) {
      setItemEditable(true);
    } else {
      setItemId(id);
      setItemName(name);
      setItemDescription(description);
      setItemCreatedDate(createdDate);
    }
  }, []);

  const resetForm = () => {
    setItemId('');
    setItemName('');
    setItemDescription('');
    setItemCreatedDate('');
  };

  const createItem = (data, items) => {
    const newListItems = JSON.stringify([...items, data]);

    localStorage.setItem(TODO_LIST_ITEMS, newListItems);

    resetForm();
  };

  const updateItem = (index, data, items) => {
    const newItems = items;

    newItems[index] = data;

    const newListItems = JSON.stringify(newItems);

    localStorage.setItem(TODO_LIST_ITEMS, newListItems);

    // TODO: might be able to remove after redux implementation
    setItemEditable(false);
  };

  const handleSubmit = e => {
    e.preventDefault();

    const persistedItems = JSON.parse(localStorage.getItem(TODO_LIST_ITEMS));
    const existingIndex = persistedItems.findIndex(
      item => item.itemId === itemId
    );

    const formData = {
      itemId: itemId || shortid.generate(),
      itemName,
      itemDescription,
      itemCreatedDate: itemCreatedDate || new Date().toString()
    };

    if (existingIndex >= 0) {
      updateItem(existingIndex, formData, persistedItems);
    } else {
      createItem(formData, persistedItems);
    }
  };

  const handleOnEditClicked = () => {
    setItemEditable(true);
  };

  const handleOnDeleteClicked = () => {
    const persistedItems = JSON.parse(localStorage.getItem(TODO_LIST_ITEMS));
    const newListItems = JSON.stringify(
      persistedItems.filter(item => item.itemId !== itemId)
    );

    localStorage.setItem(TODO_LIST_ITEMS, newListItems);
  };

  const itemClass = `list-item
    ${itemEditable && 'list-item--editable'}
    ${creation && 'list-item--creation'}`;

  const showDescription =
    creation || itemEditable || (!itemEditable && itemDescription);

  return (
    <li className={itemClass}>
      <form className="list-item__form" onSubmit={handleSubmit}>
        <input
          className="list-item__form__name"
          id="name"
          type="text"
          value={itemName}
          placeholder="Add a to-do..."
          onChange={e => setItemName(e.target.value)}
          disabled={!itemEditable}
        />
        {showDescription && (
          <textarea
            className="list-item__form__description"
            id="description"
            type="text"
            value={itemDescription}
            placeholder="Give it a description... (optional)"
            onChange={e => setItemDescription(e.target.value)}
            disabled={!itemEditable}
          />
        )}
        {itemEditable ? (
          <button
            className="list-item__form__button list-item__form__button--submit"
            type="submit"
            disabled={!itemName}
          >
            {creation ? 'Create' : 'Update'}
          </button>
        ) : (
          <div className="list-item__form__bottom-wrapper">
            <p className="list-item__form__created">
              Created: {itemCreatedDate}
            </p>
            <div className="list-item__form__buttons-wrapper">
              <button
                className="list-item__form__button list-item__form__button--edit"
                type="button"
                onClick={handleOnEditClicked}
              >
                Edit
              </button>
              <button
                className="list-item__form__button list-item__form__button--delete"
                type="button"
                onClick={handleOnDeleteClicked}
              >
                Delete
              </button>
            </div>
          </div>
        )}
      </form>
    </li>
  );
};

ListItem.propTypes = {
  creation: PropTypes.bool,
  editable: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string,
  description: PropTypes.string,
  createdDate: PropTypes.string
};

ListItem.defaultProps = {
  creation: false,
  editable: false,
  id: '',
  name: '',
  description: '',
  createdDate: ''
};

export default ListItem;
