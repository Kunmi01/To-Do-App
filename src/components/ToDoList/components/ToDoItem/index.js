import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import './styles.scss';
import { TODO_LIST_ITEMS } from '../../../../utils/constants';

const shortid = require('shortid');

const ToDoItem = ({
  editable,
  creation,
  id,
  name,
  description,
  createdDate
}) => {
  const [itemEditable, setItemEditable] = useState(false);
  const [itemId, setItemId] = useState('');
  const [itemName, setItemName] = useState('');
  const [itemDescription, setItemDescription] = useState('');
  const [itemCreatedDate, setItemCreatedDate] = useState('');

  useEffect(() => {
    if (editable) {
      setItemEditable(true);
    } else {
      setItemId(id);
      setItemName(name);
      setItemDescription(description);
      setItemCreatedDate(createdDate);
    }
  }, []);

  const resetForm = () => {
    setItemId('');
    setItemName('');
    setItemDescription('');
    setItemCreatedDate('');
  };

  const createItem = (data, items) => {
    const newToDoItems = JSON.stringify([...items, data]);

    localStorage.setItem(TODO_LIST_ITEMS, newToDoItems);

    resetForm();
  };

  const updateItem = (index, data, items) => {
    const newItems = items;

    newItems[index] = data;

    const newToDoItems = JSON.stringify(newItems);

    localStorage.setItem(TODO_LIST_ITEMS, newToDoItems);

    // TODO: might be able to remove after redux implementation
    setItemEditable(false);
  };

  const deleteItem = (todoId, items) => {
    const newToDoItems = JSON.stringify(
      items.filter(item => item.itemId !== todoId)
    );

    localStorage.setItem(TODO_LIST_ITEMS, newToDoItems);
  };

  const handleSubmit = e => {
    e.preventDefault();

    const persistedItems = JSON.parse(localStorage.getItem(TODO_LIST_ITEMS));
    const existingIndex = persistedItems.findIndex(
      item => item.itemId === itemId
    );

    const formData = {
      itemId: itemId || shortid.generate(),
      itemName,
      itemDescription,
      itemCreatedDate: itemCreatedDate || new Date().toString()
    };

    if (existingIndex >= 0) {
      updateItem(existingIndex, formData, persistedItems);
    } else {
      createItem(formData, persistedItems);
    }
  };

  const handleOnEditClicked = () => {
    setItemEditable(true);
  };

  const handleOnDeleteClicked = () => {
    const persistedItems = JSON.parse(localStorage.getItem(TODO_LIST_ITEMS));
    deleteItem(itemId, persistedItems);
  };

  const todoItemClass = `todo-item
    ${itemEditable && 'todo-item--editable'}
    ${creation && 'todo-item--creation'}`;

  const showDescription =
    creation || itemEditable || (!itemEditable && itemDescription);

  return (
    <li className={todoItemClass}>
      <form className="todo-item__form" onSubmit={handleSubmit}>
        <input
          className="todo-item__form__name"
          id="name"
          type="text"
          value={itemName}
          placeholder="Add a to-do..."
          onChange={e => setItemName(e.target.value)}
          disabled={!itemEditable}
        />
        {showDescription && (
          <textarea
            className="todo-item__form__description"
            id="description"
            type="text"
            value={itemDescription}
            placeholder="Give it a description... (optional)"
            onChange={e => setItemDescription(e.target.value)}
            disabled={!itemEditable}
          />
        )}
        {itemEditable ? (
          <button
            className="todo-item__form__button todo-item__form__button--submit"
            type="submit"
            disabled={!itemName}
          >
            {creation ? 'Create' : 'Update'}
          </button>
        ) : (
          <div className="todo-item__form__bottom-wrapper">
            <p className="todo-item__form__created">
              Created: {itemCreatedDate}
            </p>
            <div className="todo-item__form__buttons-wrapper">
              <button
                className="todo-item__form__button todo-item__form__button--edit"
                type="button"
                onClick={handleOnEditClicked}
              >
                Edit
              </button>
              <button
                className="todo-item__form__button todo-item__form__button--delete"
                type="button"
                onClick={handleOnDeleteClicked}
              >
                Delete
              </button>
            </div>
          </div>
        )}
      </form>
    </li>
  );
};

ToDoItem.propTypes = {
  creation: PropTypes.bool,
  editable: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string,
  description: PropTypes.string,
  createdDate: PropTypes.string
};

ToDoItem.defaultProps = {
  creation: false,
  editable: false,
  id: '',
  name: '',
  description: '',
  createdDate: ''
};

export default ToDoItem;
