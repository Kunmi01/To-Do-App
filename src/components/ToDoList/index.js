import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import ToDoItemContainer from './components/ToDoItem';
import './styles.scss';

const mapStateToProps = state => {
  return { toDoList: state.toDoList };
};

export const ToDoList = ({ toDoList }) => (
  <div className="todo-list">
    <div className="todo-list__wrapper">
      <h3 className="todo-list__header">To-do List</h3>
      <ul className="todo-list__inner-wrapper">
        <ToDoItemContainer creation editable />
        {!!toDoList.length &&
          toDoList
            .map(item => (
              <ToDoItemContainer
                key={item.itemId}
                id={item.itemId}
                name={item.itemName}
                description={item.itemDescription}
                createdDate={item.itemCreatedDate}
              />
            ))
            .reverse()}
      </ul>
    </div>
  </div>
);

ToDoList.propTypes = {
  toDoList: PropTypes.arrayOf(
    PropTypes.shape({
      itemId: PropTypes.string.isRequired,
      itemName: PropTypes.string.isRequired,
      itemDescription: PropTypes.string,
      itemCreatedDate: PropTypes.string.isRequired
    })
  ).isRequired
};

export default connect(mapStateToProps)(ToDoList);
