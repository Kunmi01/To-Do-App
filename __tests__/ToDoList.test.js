import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import ConnectedToDoList, { ToDoList } from '../src/components/ToDoList';
import { ToDoItem } from '../src/components/ToDoList/components/ToDoItem';

describe('ToDoList', () => {
  const initialState = {
    toDoList: []
  };
  const mockStore = configureStore();
  let store;
  let container;
  let toDoListContainer;

  beforeEach(() => {
    store = mockStore(initialState);
    container = mount(
      <Provider store={store}>
        <ConnectedToDoList />
      </Provider>
    );
    toDoListContainer = container.find(ToDoList);
  });

  it('renders connected(SMART) component', () => {
    expect(container.find(ConnectedToDoList).exists()).toBe(true);
    expect(toJson(container)).toMatchSnapshot();
  });

  it('prop match with initialState', () => {
    expect(toDoListContainer.prop('toDoList')).toEqual(initialState.toDoList);
  });

  it('contains 1 creation ToDoItem', () => {
    expect(toDoListContainer.find(ToDoItem)).toHaveLength(1);
    expect(toDoListContainer.find(ToDoItem).prop('creation')).toBe(true);
    expect(toDoListContainer.find(ToDoItem).prop('editable')).toBe(true);
  });
});
