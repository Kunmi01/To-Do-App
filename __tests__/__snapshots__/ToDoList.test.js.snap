// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList renders connected(SMART) component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ToDoList)>
    <ToDoList
      dispatch={[Function]}
      toDoList={Array []}
    >
      <div
        className="todo-list"
      >
        <div
          className="todo-list__wrapper"
        >
          <h3
            className="todo-list__header"
          >
            To-do List
          </h3>
          <ul
            className="todo-list__inner-wrapper"
          >
            <Connect(ToDoItem)
              creation={true}
              editable={true}
            >
              <ToDoItem
                createdDate=""
                creation={true}
                description=""
                dispatch={[Function]}
                editable={true}
                id=""
                name=""
              >
                <li
                  className="todo-item todo-item--editable todo-item--creation"
                >
                  <form
                    className="todo-item__form"
                    onSubmit={[Function]}
                  >
                    <input
                      className="todo-item__form__name"
                      disabled={false}
                      id="name"
                      onChange={[Function]}
                      placeholder="Add a to-do..."
                      type="text"
                      value=""
                    />
                    <textarea
                      className="todo-item__form__description"
                      disabled={false}
                      id="description"
                      onChange={[Function]}
                      placeholder="Give it a description... (optional)"
                      type="text"
                      value=""
                    />
                    <button
                      className="todo-item__form__button todo-item__form__button--submit"
                      disabled={true}
                      type="submit"
                    >
                      Create
                    </button>
                  </form>
                </li>
              </ToDoItem>
            </Connect(ToDoItem)>
          </ul>
        </div>
      </div>
    </ToDoList>
  </Connect(ToDoList)>
</Provider>
`;
